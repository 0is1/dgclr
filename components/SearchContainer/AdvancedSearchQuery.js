// @flow
import React, { Component } from 'react';
import gql from 'graphql-tag';
import { graphql } from 'react-apollo';
import Link from 'next/link';
import { connect } from 'react-redux';
import { differenceBy, size } from 'lodash';
import {
  setCourses as setCoursesFunc,
  setAdvancedSearchQuery,
} from 'components/SearchContainer/actions';
import {
  latestAdvancedQuery as latestQueryFunc,
  queryResultsFromState as queryResultsFromStateFunc,
} from 'components/SearchContainer/selectors';
import { getRandomKey, uniqueLayoutRatings } from 'helpers/utils';
import LayoutRatingBadges from 'components/Layout/Badges';
import { ClipLoader } from 'components/Spinners';
import Styles from 'components/SearchContainer/SearchContainer.styles';
import BaseStyles from 'components/Container/Container.styles';
import type { GraphQLData, State } from 'lib/types';

const { UL, LI } = BaseStyles;

const { SearchResultItem } = Styles;

type Props = {
  filter: {},
  queryResults: [],
  data: GraphQLData,
  latestQuery: string,
  setCourses: Function,
  setSearchQuery: Function,
};

class AdvancedSearchQuery extends Component<Props> {
  componentDidUpdate(prevProps, prevState, snapshot) {
    if (snapshot !== null) {
      const { data = {} } = this.props;
      const { courses = [] } = data;
      if (courses) {
        this.setCourses(courses);
      }
    }
  }

  getSnapshotBeforeUpdate(prevProps) {
    const { data = {}, latestQuery, filter } = this.props;
    const { courses = [] } = data;
    const prevCourseData = prevProps.data;
    const prevCourseArr = prevCourseData && prevProps.data.courses;
    const prevCourses = prevCourseArr && prevCourseArr.length ? prevCourseArr : [];
    const stringifyFilter = JSON.stringify(filter);
    // console.log('courses: ', courses);
    // console.log('prevCourses: ', prevCourses);
    if (
      (courses && courses.length && differenceBy(courses, prevCourses, '_id').length > 0)
      || (!courses.length && stringifyFilter !== latestQuery)
      || (courses.length !== prevCourses.length && stringifyFilter !== latestQuery)
    ) {
      return true;
    }
    return null;
  }

  setCourses = (courses) => {
    const { setCourses, setSearchQuery, filter } = this.props;
    setCourses(courses);
    setSearchQuery(courses, JSON.stringify(filter));
  };

  render() {
    const { filter, queryResults, data = {} } = this.props;
    if (size(filter) < 1 && !queryResults.length) return null;
    if (data && data.loading) {
      return <ClipLoader />;
    }
    const { courses } = data;
    const courseData = queryResults.length ? queryResults : courses;
    let results = <li>Ei hakutuloksia!</li>;
    if (courseData && courseData.length) {
      results = courseData.map((course) => {
        const ratings = uniqueLayoutRatings(course.layouts);
        return (
          <LI key={getRandomKey()}>
            <Link as={`/${course.slug}`} href={`/course?slug=${course.slug}`}>
              <SearchResultItem>
                {course.name}
                <LayoutRatingBadges tiny ratings={ratings} />
              </SearchResultItem>
            </Link>
          </LI>
        );
      });
    }
    return <UL>{results}</UL>;
  }
}

const mapStateToProps = (state: State) => ({
  latestQuery: latestQueryFunc(state),
  queryResults: queryResultsFromStateFunc(state, latestQueryFunc(state)),
});

const mapDispatchToProps = dispatch => ({
  setCourses: courses => dispatch(setCoursesFunc(courses)),
  setSearchQuery: (courses, query) => dispatch(setAdvancedSearchQuery(courses, query)),
});

const SEARCH_COURSES = gql`
  query AdvancedCoursesQuery($filter: CourseQueryFilterInput!) {
    courses(filter: $filter) {
      _id
      name
      description
      slug
      courseInfo {
        basketType
        teeType
        infoSignType
        surfaceShapeTypes
        courseTypes
        mapUrl
        founded
        maintenanceCycle
        rangeMaster
        courseDesigner
        fee {
          amount
          currency
          value
        }
      }
      locationInfo {
        address
        city
        zip
        location {
          coordinates
        }
      }
      layouts {
        name
        rating
        holes {
          par
          length {
            foot
            meter
          }
        }
        holeCount
        totalPar
      }
    }
  }
`;

// The `graphql` wrapper executes a GraphQL query and makes the results
// available on the `data` prop of the wrapped component (PostList)
const ComponentWithMutation = graphql(SEARCH_COURSES, {
  options: ({ filter }) => ({
    variables: {
      filter,
    },
  }),
  props: ({ data }) => ({
    data,
  }),
  skip: ({ filter }) => !filter || size(filter) < 1,
})(AdvancedSearchQuery);

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(ComponentWithMutation);
