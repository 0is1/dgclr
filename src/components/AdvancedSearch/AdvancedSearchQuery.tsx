import { useQuery } from '@tanstack/react-query';
import { Space, Table, Tag } from 'antd';
import { useRouter } from 'next/router';
// import type { SortOrder } from 'antd/lib/table/interface';
// import Link from 'next/link';
import useLocalStorageState from 'use-local-storage-state';
import { useTranslation } from 'next-i18next';
// import { Course } from '../../types';

import { getCourses } from '../../graphql/fetcher';
import { CourseQueryFilterInput } from '../../types';
import AdvancedMapComponent from '../Maps/AdvancedMapComponent';

function AdvancedSearchQuery() {
  const { t } = useTranslation(['common']);
  const router = useRouter();
  const [query, setQuery] = useLocalStorageState('advanced_search', {
    defaultValue: {
      filter: {},
      limit: 50,
      events: {},
      skip: 0,
    },
  });
  console.log('query: ', query);
  if (query.limit !== 50) {
    setQuery({
      ...query,
      limit: 50,
    });
  }
  const { data, failureReason, isError, isLoading } = useQuery({
    queryKey: [`courses_${JSON.stringify(query)}`],
    queryFn: async () => {
      const { filter, limit } = query;
      const filterAsFilterInput = filter as CourseQueryFilterInput;
      const data = await getCourses({
        filter: filterAsFilterInput,
        limit,
      });
      return data;
    },
    retry: 2,
  });
  if (isError) {
    console.log(failureReason);
    return (
      <div>
        <>{t('common:search_error')}</>
      </div>
    );
  }
  if (isLoading) {
    return <h1>Loading</h1>;
  }
  console.log('data: ', data);
  const locations = data?.courses.map((course) => {
    const { locationInfo } = course;
    const { location } = locationInfo;
    const { coordinates } = location;
    if (!coordinates) {
      return {
        id: course._id,
        name: course.name,
        slug: course.slug,
        lat: undefined,
        lng: undefined,
      };
    }
    return {
      id: course._id,
      name: course.name,
      slug: course.slug,
      lat: coordinates[1],
      lng: coordinates[0],
    };
  });
  return <AdvancedMapComponent locations={locations} />;
}

export default AdvancedSearchQuery;
