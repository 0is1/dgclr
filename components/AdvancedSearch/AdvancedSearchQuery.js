// @flow
import React, { Component } from 'react';
import gql from 'graphql-tag';
import { graphql } from 'react-apollo';
import Link from 'next/link';
import { connect } from 'react-redux';
import { differenceBy, size } from 'lodash';
import { Box } from 'rebass';
import { GoChevronRight } from 'react-icons/go';
import { getRandomKey, uniqueLayoutRatings } from 'helpers/utils';
import { COURSE_QUERY } from 'lib/constants';
import LayoutRatingBadges from 'components/Layout/Badges';
import { ClipLoader } from 'components/Spinners';
import Styles from 'components/SearchContainer/SearchContainer.styles';
import BaseStyles from 'components/Container/Container.styles';
import type { GraphQLData, State } from 'lib/types';
import { setCourses as setCoursesFunc } from 'components/Course/actions';
import { queryResultsFromState as queryResultsFromStateFunc } from 'components/SearchContainer/selectors';
import { setAdvancedSearchQuery } from './actions';
import { latestAdvancedQuery as latestQueryFunc } from './selectors';
import AdvancedSearchQueryStyles from './AdvancedSearchQuery.styles';

const { UL, LI } = BaseStyles;
const { SearchResultItem, SearchResultIcon } = Styles;
const { NoResults } = AdvancedSearchQueryStyles;
type Props = {
  filter: {},
  queryResults: [],
  data: GraphQLData,
  latestQuery: string,
  setCourses: Function,
  setSearchQuery: Function,
};

export class AdvancedSearchQuery extends Component<Props> {
  componentDidUpdate(prevProps, prevState, snapshot) {
    if (snapshot !== null) {
      const { data = {} } = this.props;
      const { courses = [] } = data;
      if (courses) {
        this.setCourses(courses);
      }
    }
  }

  getSnapshotBeforeUpdate(prevProps) {
    const { data = {}, latestQuery, filter } = this.props;
    const { courses = [] } = data;
    const prevCourseData = prevProps.data;
    const prevCourseArr = prevCourseData && prevProps.data.courses;
    const prevCourses = prevCourseArr && prevCourseArr.length ? prevCourseArr : [];
    const stringifyFilter = JSON.stringify(filter);
    // console.log('courses: ', courses);
    // console.log('prevCourses: ', prevCourses);
    if (
      (courses
        && courses.length
        && differenceBy(courses, prevCourses, '_id').length > 0)
      || (!courses.length && stringifyFilter !== latestQuery)
      || (courses.length !== prevCourses.length && stringifyFilter !== latestQuery)
    ) {
      return true;
    }
    return null;
  }

  setCourses = (courses) => {
    const { setCourses, setSearchQuery, filter } = this.props;
    setCourses(courses);
    setSearchQuery(courses, JSON.stringify(filter));
  };

  render() {
    const { filter, queryResults, data = {} } = this.props;
    if (size(filter) < 1 && !queryResults.length) return null;
    if (data && data.loading) {
      return <ClipLoader />;
    }
    const { courses } = data;
    const courseData = queryResults.length ? queryResults : courses;
    let results = false;
    if (courseData && courseData.length) {
      results = courseData.map((course) => {
        const ratings = uniqueLayoutRatings(course.layouts);
        return (
          <LI key={getRandomKey()} listStyle="none">
            <Link as={`/${course.slug}`} href={`/course?slug=${course.slug}`}>
              <SearchResultItem>
                {course.name}
                <LayoutRatingBadges tiny ratings={ratings} />
                <SearchResultIcon>
                  <GoChevronRight size="1.5rem" color="#0067ee" />
                </SearchResultIcon>
              </SearchResultItem>
            </Link>
          </LI>
        );
      });
    }
    if (!results) {
      return (
        <Box p={[0, '0.5rem 2rem']}>
          <NoResults>Ei hakutuloksia!</NoResults>
        </Box>
      );
    }
    return (
      <Box p={[0, '0.5rem 2rem']}>
        <UL>{results}</UL>
      </Box>
    );
  }
}

const mapStateToProps = (state: State) => ({
  latestQuery: latestQueryFunc(state),
  queryResults: queryResultsFromStateFunc(state, latestQueryFunc(state)),
});

const mapDispatchToProps = (dispatch: Function) => ({
  setCourses: courses => dispatch(setCoursesFunc(courses)),
  setSearchQuery: (courses, query) => dispatch(setAdvancedSearchQuery(courses, query)),
});

const SEARCH_COURSES = gql`
  query AdvancedCoursesQuery($filter: CourseQueryFilterInput!) {
    courses(filter: $filter, limit: 150) {
      ${COURSE_QUERY}
    }
  }
`;

// The `graphql` wrapper executes a GraphQL query and makes the results
// available on the `data` prop of the wrapped component (PostList)
const ComponentWithMutation = graphql(SEARCH_COURSES, {
  options: ({ filter }) => ({
    variables: {
      filter,
    },
  }),
  props: ({ data }) => ({
    data,
  }),
  skip: ({ filter }) => !filter || size(filter) < 1,
})(AdvancedSearchQuery);

export default connect<any, Props, any, any, any, Function>(
  mapStateToProps,
  mapDispatchToProps,
)(ComponentWithMutation);
