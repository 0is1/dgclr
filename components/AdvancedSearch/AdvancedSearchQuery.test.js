import React from 'react';
import { shallow } from 'enzyme';
import { AdvancedSearchQuery } from './AdvancedSearchQuery';

const mockProps = (props = {}) => ({
  filter: {},
  queryResults: [],
  data: [],
  latestQuery: '',
  setCourses: jest.fn(),
  setSearchQuery: jest.fn(),
  ...props,
});

describe('<AdvancedSearchQuery />', () => {
  describe('snapshots', () => {
    it('renders with default props', () => {
      const props = mockProps();
      const subject = shallow(<AdvancedSearchQuery {...props} />);
      expect(subject).toMatchSnapshot();
    });
  });
});
