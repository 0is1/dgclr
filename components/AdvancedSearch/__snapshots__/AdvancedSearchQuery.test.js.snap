// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AdvancedSearchQuery /> snapshots renders with default props 1`] = `""`;

exports[`<AdvancedSearchQuery /> snapshots renders with queryResults 1`] = `
<Box
  p={
    Array [
      0,
      "0.5rem 2rem",
    ]
  }
>
  <Containerstyles__UL>
    <Containerstyles__LI
      key="44200-suolahti-äänekoski"
      listStyle="none"
    >
      <Link
        as="/44200-suolahti-äänekoski"
        href="/course?slug=44200-suolahti-äänekoski"
      >
        <SearchContainerstyles__SearchResultItem>
          44200 Suolahti
          <LayoutRatingBadges
            ratings={
              Array [
                "A1",
              ]
            }
            tiny={true}
          />
          <SearchContainerstyles__SearchResultIcon>
            <GoChevronRight
              color="#0067ee"
              size="1.5rem"
            />
          </SearchContainerstyles__SearchResultIcon>
        </SearchContainerstyles__SearchResultItem>
      </Link>
    </Containerstyles__LI>
    <Containerstyles__LI
      key="ford-discgolfpark-tuusula"
      listStyle="none"
    >
      <Link
        as="/ford-discgolfpark-tuusula"
        href="/course?slug=ford-discgolfpark-tuusula"
      >
        <SearchContainerstyles__SearchResultItem>
          Ford DiscGolfPark
          <LayoutRatingBadges
            ratings={
              Array [
                "AA1",
              ]
            }
            tiny={true}
          />
          <SearchContainerstyles__SearchResultIcon>
            <GoChevronRight
              color="#0067ee"
              size="1.5rem"
            />
          </SearchContainerstyles__SearchResultIcon>
        </SearchContainerstyles__SearchResultItem>
      </Link>
    </Containerstyles__LI>
  </Containerstyles__UL>
</Box>
`;
